Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/Gemfile

```diff
@@ -0,0 +1,56 @@
+source 'https://rubygems.org'
```

Comment:
what is this file for?

Reply:
`Gemfile` denotes the source url to fetch the gems package and also the gems used in the project with their respective versions.

Reply:
1. `Gemfile denotes the source url` elaborate more on this pls
2. But not everything has version number there o. How do we set the respective version?
3. Why we use this?

Reply:
1. We put the source url: 'https://rubygems.org' so it will know which ruby server for it to fetch for the gem packages. Other than that, we can also wrap other sources in a block as below:
```
source 'https://rubygems.org'
# Gems from default source

source 'https://example.com' do
     # Gems from other source
end
```

2. When no version is specified next to the gem, then it will be updated to the newest version during `bundle update`. Below is an example of gem packages listed in a Gemfile.
```
gem 'nokogiri'
gem 'rails', '5.0.0'
gem 'rack',  '>=1.0'
gem 'thin',  '~>1.1'
```
`gem 'rails', '5.0.0'` means that we want `rails` to be specifically version 5.0.0 only.
`gem 'rack',  '>=1.0'` means that any version higher than or equal to 1.0 would work.
`gem 'thin',  '~>1.1'` means that any version higher than or equal to 1.1 would work but not 2.0 or higher. This is the same as `gem 'thin',  '>=1.1', '<2.0'`.

3. By having versions specified or using pessimistic locking `~>`, it ensures that when running `bundle update` no accidental break of code will happen.